\documentclass[nogrid]{MBE}%

\usepackage{url}

\jshort{mst}

\volname{}

\jvolume{0}

\jvol{}

\jissue{0}

\pubyear{2016}

\mstype{Letter}

\artid{012}

\access{Advance Access publication November 2, 2016}


\begin{document}

\title[RD-MCL]{High resolution classification of orthogroups by recursive dynamic Markov clustering}


\author[Bond et al.]{Stephen R. \surname{Bond}, Karl E.
Keat, and Andreas D. Baxevanis$^{\ast}$}

\address{Computational and Statistical Genomics Branch, Division of Intramural Research, National Human Genome Research Institute, National Institutes of Health, 50 South Drive, Bethesda, MD, USA, 20892}


\history{Received 12 November 2016}

\coresp{E-mail: andy@mail.nih.gov}


\editor{}

\abstract{}

\keyword{orthogroup, clustering}


\maketitle


\section{{Introduction}\label{sec:Intro}}
When a gene evolves an important physiological role, purifying selection tends to maintain that function through evolutionary time \cite{Altenhoff:2012ea, Rogozin:2014fp, KryuchkovaMostacci:2016iw}. As a result, orthology (i.e., homology via speciation) has become a widely used predictor of shared gene product function among species, with considerable effort made to develop computational methods for identifying orthologs. The algorithms in current use generally fall into two distinct categories: Tree-based and graph-based clustering methods \cite{Tekaia:2016ga}. Tree-based approaches (e.g., Ensembl Compara \cite{Vilella:2009ju}, LOFT \cite{vanderHeijden:2007bo}, and SYNERGY \cite{Wapinski:2007fa}) broadly rely on estimating a phylogenetic tree for a target gene family, and then reconciling the gene tree with a `known' species tree to identify orthologous clades. While tree-based methods are very accurate under ideal conditions, estimating the species trees they rely creates a considerable source of uncertainty \cite{Xu:2016ek}. Alternatively, pairwise similarity graph clustering methods leverage graph theory to rapidly identify groups of related sequences from genome scale datasets [REF]. Reciprocal best-hit methods were among the earliest developed for this purpose, but were restricted to assessing only two species at a time [REF]. Due to the non-transitive nature of orthology (i.e., paralogs in one species can be orthologous to a single gene in another species), it is more difficult (or impossible) to explicitly assign sequences into groups of pure orthologs [REF]. Instead, the term `orthogroup' has come to represent a cluster of orthologs that may include closely related paralogs [REF]. InParanoid, EggNOG \cite{Jensen:2007cc}, and OMA \cite{Roth:2009iu} are popular tools for assigning sequences to orthogroups using a `best-hit clique' approach, where closed best-hit sub-graphs are identified in the dataset. While accurate within each sub-graph, these methods tend to be overly strict in their assignment; this causes an under-representation of actual orthologous relationships among many species. Alternatively, Markov clustering (MCL) is very efficient at isolating more inclusive sub-graphs. OrthoMCL is one of the most popular MCL-based ortholog prediction methods, but it is prone to placing too many in-paralogs into orthogroups (i.e., it is less precise). In the current study we have increased the overall resolving power of de novo MCL-based orthogroup assignment with a number of novel enhancements, including refinement of the pairwise similarity metrics, using a supervised heuristic to dynamically select MCL parameters, recursively subdividing orthogroups, and testing putative orthogroups for best-hit cliques to maximize resolution.

BLAST scores (bit or e-value) have a strong length bias when calculating orthogroups \cite{Emms:2015ig}.

\begin{table}[!t]
\tableparts{\caption{File format support provided by each BuddySuite module for reading (R) and writing (W).\label{table:formats}}}
      {\tabcolsep=4pt\begin{tabular}{@{\extracolsep{\fill}}llll@{}}
        \toprule
        Format							& SeqBuddy  					& AlignBuddy   					& PhyloBuddy
        \\\colrule
        Clustal 						& R \& W\textsuperscript{\dag} 	& R \& W						& None \\ 
        EMBL\textsuperscript{\ddag} 	& R \& W						& R\textsuperscript{\dag}\& W	& None
        \\\botrule
      \end{tabular}}
{\tablenote{\textsuperscript{\dag}All sequences must be the same length \\
      		  \textsuperscript{\ddag}Supports rich sequence annotation}}
\end{table}


\section{Methods}
I used MAFFT \cite{Katoh:2013hm}, because it's awesome.

The COG, KOG, arCOG databases may all be rich sources of data for validation. COGs are `clusters of orthologous genes', which can includes many individual orthogroups.

KEGG OCs may work in place of KOGs \cite{Nakaya:2013gg}

Construct species trees with *BEAST (Heled and Drummond 2010) and BPP (Yang and Rannala 2014; Rannala and Yang 2016) to do a tree based comparison against RD-MCL.

Possible sample data: CYP proteins \cite{Pan:2016jg}

Might want to compare results agains Ortholog-Finder if appropriate \cite{Horiike:2016dq}

Construction of an ortholog ontology \cite{Chiba:2015ed}?

Try OrthoFinder length-normalized bit scores as similarity metric between sequences \cite{Emms:2015ig}.

\section{Results}

\section{Discussion}
Spill some ink regarding in/out paralogs \cite{Sonnhammer:2002vm,Tekaia:2016ga}.

\section{Conclusions}

\section{Acknowledgments}
This research was supported by the Intramural Research Program of the National Human Genome Research Institute, National Institutes of Health.



%\begin{figure*}[t]
%\begin{center}
%\includegraphics[height=0.4\textheight]{figures/figure_fixed.eps}
%\end{center}
%\caption{RefSeq records were identified in GenBank using the following query: ``Nematoda''[Organism] AND biomol\_mrna[PROP] AND refseq[filter]. The results were downloaded in GenBank format and subsamples of 10, 100, 1000, and 10000 records were used to test the runtime performance of the BuddySuite tools (excluding tools that depend on third-party programs or services). Each dot denotes a single BuddySuite tool, runtimes are the average of 10 replicates expressed in seconds (the y-axis is log-scale), and the subscript numbers below each jitter plot represents the size of the sequence, alignment, or tree file in bytes.}
%\label{fig:timeit}
%\end{figure*}


\bibliographystyle{natbib}  % natbib.sty
\bibliography{../references/refs}         % refs.bib

\end{document}