\documentclass[twocolumn]{bmcart}  % uncomment this for twocolumn layout and comment line below
%% BioMed_Central_Tex_Template_v1.06
%%                                      %
%  bmc_article.tex            ver: 1.06 %
%                                       %

%%IMPORTANT: do not delete the first line of this template
%%It must be present to enable the BMC Submission system to
%%recognise this template!!

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%          <8 June 2012>              %%
%%                                     %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.html and the instructions for   %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%% additional documentclass options:
%  [doublespacing]
%  [linenumbers]   - put the line numbers on margins

%%% loading packages, author definitions

%\documentclass[twocolumn]{bmcart}% uncomment this for twocolumn layout and comment line below
%\documentclass{bmcart}

%%% Load packages
\usepackage{amsthm,amsmath}
%\RequirePackage{natbib}
%\RequirePackage[authoryear]{natbib}% uncomment this for author-year bibliography
%\RequirePackage{hyperref}
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails
\usepackage{graphicx}
\usepackage{lipsum}
\usepackage{textgreek}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %%
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %%
%%  submitted article.                         %%
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%\def\includegraphic{}
%\def\includegraphics{}



%%% Put your definitions there:
\startlocaldefs
\endlocaldefs


%%% Begin ...
\begin{document}

%%% Start of article front matter
\begin{frontmatter}

\begin{fmbox}
\dochead{Software}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the title of your article here     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\title{Recursive Dynamic Markov Clustering for fine-grained orthogroup classification}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors here                   %%
%%                                          %%
%% Specify information, if available,       %%
%% in the form:                             %%
%%   <key>={<id1>,<id2>}                    %%
%%   <key>=                                 %%
%% Comment or delete the keys which are     %%
%% not used. Repeat \author command as much %%
%% as required.                             %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\author[
   addressref={aff1},                    % id's of addresses, e.g. {aff1,aff2}
   %noteref={n1},                        % id's of article notes, if any
   %corref={aff1},                       % id of corresponding address, if any
   %email={steve.bond@nih.gov}            % email address
]{\inits{SR}\fnm{Stephen R} \snm{Bond}}
\author[
   addressref={aff1},
   %email={karlkeat@gmail.com}
]{\inits{P}\fnm{Paul} \snm{Gonzalez}}
\author[
   addressref={aff1},
   %email={karlkeat@gmail.com}
]{\inits{KE}\fnm{Karl E} \snm{Keat}}
\author[
   addressref={aff1},
   email={andy@mail.nih.gov}
]{\inits{AD}\fnm{Andreas D} \snm{Baxevanis}}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors' addresses here        %%
%%                                          %%
%% Repeat \address commands as much as      %%
%% required.                                %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\address[id=aff1]{%                           % unique id
  \orgname{Computational and Statistical
   Genomics Branch, Division of Intramural    % university, etc
    Research, National Human Genome
     Research Institute, National
      Institutes of Health},
  \street{50 South Drive},                     %
  \postcode{20892}                             % post or zip code
  \city{Bethesda},                             % city
  \state{MD},
  \cny{USA}                                    % country
}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter short notes here                   %%
%%                                          %%
%% Short notes will be after addresses      %%
%% on first page.                           %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%\begin{artnotes}
%\note{Sample of title note}     % note to the article
%\note[id=n1]{Equal contributor} % note, connected to author
%\end{artnotes}

%\end{fmbox}% comment this for two column layout

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Abstract begins here                 %%
%%                                          %%
%% Please refer to the Instructions for     %%
%% authors on http://www.biomedcentral.com  %%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{abstractbox}

\begin{abstract} % abstract
\parttitle{Background}  % Required
Blahh

\parttitle{Results}  % Required
Blahh

\parttitle{Conclusions}  % Required
Blahh

\end{abstract}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The keywords begin here                  %%
%%                                          %%
%% Put each keyword in separate \kwd{}.     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{keyword}  % Three to ten keywords
\kwd{orthogroup}
\kwd{ortholog}
\kwd{Markov clustering}
\end{keyword}

% MSC classifications codes, if any
%\begin{keyword}[class=AMS]
%\kwd[Primary ]{}
%\kwd{}
%\kwd[; secondary ]{}
%\end{keyword}

\end{abstractbox}
%
\end{fmbox} % uncomment this for twcolumn layout

\end{frontmatter}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Main Body begins here                %%
%%                                          %%
%% Please refer to the instructions for     %%
%% authors on:                              %%
%% http://www.biomedcentral.com/info/authors%%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%% See the Results and Discussion section   %%
%% for details on how to create sub-sections%%
%%                                          %%
%% use~\cite{...} to cite references        %%
%% ~\cite{koon} and                         %%
%% ~\cite{oreg,khar,zvai,xjon,schn,pond}    %%
%%  \nocite{smith,marg,hunn,advi,koha,mouse}%%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%% start of article main body
% <put your article body there>

%%%%%%%%%%%%%%%%
%% Background %%
%%
\section{Background}\label{sec:background}
When a gene evolves an important physiological function, purifying selection tends to maintain that function through evolutionary time~\cite{Altenhoff:2012ea,KryuchkovaMostacci:2016iw,Rogozin:2014fp}.
As a result, orthology (i.e., homology via speciation) has become a widely used predictor of shared gene product function among species, and considerable effort has been made to develop computational tools for identifying orthologs from genomic data.
Due to the non-transitive nature of orthology (i.e., paralogs in one species can be orthologous to a single gene in another species), grouping pure sets of one-to-one orthologs is often not possible~\cite{Fitch:2000tf}.
Instead, the term `orthogroup' has come to represent a cluster of genes descended from a common ancestor of the clade in question, which may include paralogs~\cite{Wapinski:2007fa}.
The algorithms currently in popular use for defining orthogroups fall into three broad categories: Synteny-based, tree-based, and graph-based clustering methods (recently reviewed in~\cite{Tekaia:2016ga} and~\cite{Habermann2016}).

Syntenic neighborhoods degrade rapidly with evolutionary distance, so pure synteny-based approaches are not generally appropriate except between closely related taxa~\cite{Kristensen:2011gw}.
Furthermore, such approaches require highly contiguous genomic assemblies;
while advances in long-read sequencing technology and de novo assembly will likely allow future genome sequencing efforts to achieve the necessary standards, current genome projects often remain highly fragmented~\cite{Koren:2015il}.

Tree-based approaches (e.g., Ensembl Compara~\cite{Vilella:2009ju}, LOFT~\cite{vanderHeijden:2007bo}, and SYNERGY~\cite{Wapinski:2007fa}) identify orthologous clades by estimating phylogenetic trees for a target gene family, and then attempt to reconcile those gene trees against a `known' species tree.
The accuracy of tree-based orthology prediction methods is tied closely to the accuracy of the species trees they rely on;
this can lead to considerable uncertainty or error, especially for less well-studied taxonomic groups~\cite{Xu:2016ek}.

Finally, pairwise similarity methods leverage graph theory to rapidly identify groups of related sequences.
InParanoid~\cite{OBrien:2005cy}, EggNOG~\cite{Jensen:2007cc}, and OMA~\cite{Roth:2009iu} are popular tools for assigning sequences to orthogroups using a `best-hit clique' approach, where closed best-hit sub-graphs are identified in the dataset.
These methods can be fast and accurate for detecting one-to-one orthologs, but they suffer diminishing recall rates when in-paralogs are present among the species under study~\cite{Dalquen:2013fz} (`in-paralog' describes homologs derived from genetic duplication \textit{after} speciation~\cite{Sonnhammer:2002vm,Tekaia:2016ga}).
In contrast, Markov clustering (MCL) can efficiently isolate more inclusive sub-graphs~\cite{VanDongen:kJZ890qx,Enright:2002uq}, although the trade-off is generally a reduction in precision;
it is often difficult to separate closely related orthogroups within a given gene family.
Indeed, the major challenge faced when analyzing a defined gene family is \textit{resolution}, because the popular MCL-based ortholog prediction methods, such as OrthoMCL~\cite{Li:2003en}, OrthoFinder~\cite{Emms:2015ig}, and ProteinOrtho~\cite{Lechner:2011jk}, are targeted towards coarse-grained clustering of all protein models derived from whole-genome data.
While computationally efficient, these resources are not well suited for fine-grained processing of individual gene families where all input sequences are known to be homologous from the outset.

[!!!Need to address Kimmen Sjolander's work~\cite{Brown:2007dp,Krishnamurthy:2007fs}!!!]

This leaves a gap in our ability to easily discern evolutionary patterns at this scale, which may be exacerbating the propagation of annotation errors in our public databases~\cite{Schnoes:2009gb}, where annotation transfer following inference of homology against a limited number of reference species is largely the \textit{modus operandi}~\cite{Aken:2016dl,Mi:2016bw,OLeary:2016cm}.
Here we present a method that overcomes these challenges, along with a suite of tools to assist with detailed downstream analysis.


\section{Results and Discussion}\label{sec:resultsAndDiscussion}
Recursive Dynamic Markov Clustering (RD-MCL) is a new pairwise similarity graph-based orthogroup prediction heuristic that achieves high precision within the context of a gene family.
Specifically, RD-MCL improves upon the BLASTP-based pairwise similarity metrics currently in popular use, applies an optimization algorithm to dynamically select MCL parameters, recursively subdivides overly inclusive orthogroups, and implements final polishing steps to maximize overall accuracy (Figure~\ref{fig:pipeline}).
The entire workflow has been encapsulated in a single executable, which can be run from the command line and minimally requires a sequence file as input where each record is prefixed with a species identifier.
Given appropriate taxonomic coverage, RD-MCL clearly reveals mis-assembled or mis-annotated sequences, as well as orthogroups that have been previously undescribed.
Furthermore, a set of high quality orthogroups from a well-sampled taxonomic group can be leveraged to analyze sequences from clades that are less well sampled, allowing for detailed phylogenetic placement of new sequences into a gene family with greater precision than is possible with simple best-hit database queries.
The software is open-source (https://research.nhgri.nih.gov/software/RD-MCL/) and distributed as part of a suite of tools to facilitate all of the downstream analyses reported here.

\begin{figure*}[t]
  \begin{center}
  \includegraphics[height=0.6\textheight]{../figures/pipeline.eps}
\end{center}
\caption{Flow diagram illustrating the RD-MCL pipeline.
This entire workflow is implemented in the RD-MCL software, so the user is only responsible for providing a sequence file as input.}
\label{fig:pipeline}
\end{figure*}

\subsection{Wnt proteins as a benchmark}\label{subsec:benchmarkDatasetWnts}
The `wingless-related integration site' (Wnt) protein family contains approximately 20 members that are remarkably well conserved in vertebrates.
Because of their central role in key biological processes, like embryonic patterning and cellular proliferation, these genes have been carefully and comprehensively cataloged~\cite{Garriock:2007hb}, allowing us to leverage them as a pre-curated dataset for benchmarking purposes in this study.
The family itself is divided into twelve distinct classes (Wnt1-Wnt11, and Wnt16), many of which are subdivided into two paralogs in mammals and bird, or up to four paralogs in fish.
Primary isoforms for each Wnt gene belonging to species in \textit{Mammalia}, \textit{Aves}, or \textit{Actinopterygii} were identified in the RefSeq database using PSI-BLAST, and only minor filtering was applied to remove sequences that clearly fall outside of the normal complement of orthogroups (e.g.,\ sequences smaller than 250 or larger than 650 residues, and several duplicate entries for \textit{Homo sapiens}, \textit{Mus musculus}, \textit{Rattus norvegicus}, and \textit{Bos taurus}).
Twenty-one benchmark orthogroups were created from this dataset, strictly based on RefSeq annotation as summarized in supplemental table X\@.

\subsection{Defining an orthogroups score}\label{subsec:definingAnOrthogroupScore}
If an orthogroup is explicitly defined as ``the set of genes that are descended from a single gene in the last common ancestor of all the species being considered"~\cite{Emms:2015ig}, then we can devise a scoring system to assess the relative quality of any putative orthogroups based on the number of genes from each organism in a cluster and the overall distribution of genes across all clusters (i.e.,\ independent of sequence alignment or phylogenetic inference).
The details of this `orthogroups score' are outlined in subsection~\ref{subsec:rd-mclFitnessFunction} of the Materials and Methods and Figure~\ref{fig:orthogroupscores}, but, intuitively, a cluster with many sequences each from distinct organisms will be scored higher than a cluster of the same size with many sequences from the same organism.
Furthermore, the relative influence a given species can have on the score is scaled according to how many paralogs that species possesses, thereby preventing species-specific gene expansions from confounding the overall picture of homology within the family.
While these scores allow us to compare putative orthogroups within a given dataset, they are relative and unitless, so they should not be used to directly compare orthogroups from separate datasets.

\begin{figure*}[t]
  \begin{center}
  \includegraphics[height=0.5\textheight]{../figures/orthogroup_scores.eps}
\end{center}
\caption{Cluster scoring function.
A)~Cluster scores always increase as more sequences are added, but the rate of increase is reduced with the addition of in-paralogs.
Each series represents a square matrix of X species with X paralogs each, where sequences are added to the sample cluster one at a time, always including the minimum number of in-paralogs possible.
Note that the inflection points corresponds to the next 'level' of in-paralogs being added to the cluster.
B)~To further demonstrate the diminishing returns in-paralogs contribute to cluster scores, orthologs in fixed-size clusters are progressively replaced with in-paralogs, thereby reducing the score of that cluster.
C)~To further inflate the score of clusters with diverse composition, a scaling factor is applied according to the percentage of available taxa a cluster contains.
In the example plot, the cluster is composed of a single sequence and its score is ascertained as taxa are iteratively added to the total data set.
This scaling factor is constrained between 1 and 2.
D)~The sequences from the taxon with the greatest number of paralogs in the entire dataset are given a base score of 1, with all other sequences scaled by a factor of sqrt(len(max)/len(taxon)).
Thus, species' with fewer paralogs contribute more to the score than those with a greater number of paralogs. }
\label{fig:orthogroupscores}
\end{figure*}


\subsection{Sequence similarity metrics}\label{subsec:sequenceSimilarityMetrics}
Historically, the negative common log (-log\textsubscript{10}) of the BLASTP expectation value (NLE) has been the most widely used edge-weighting metric for graph-based ortholog prediction~\cite{Enright:2002uq,Li:2003en,Gibbons:2015be}.
The popularity of BLAST as an ortholog assessment metric is understandable, given that it is one of the most widely used and trusted tools in bioinformatics~\cite{Altschul:1990dy} and strikes a convenient balance between speed and sensitivity when working with genome-scale datasets.
Unfortunately, the common implementations of BLASTP fail to resolve differences among any sequences with similarities that return an E-value of less than 10\textsuperscript{-180}, which can become problematic if trying to separate closely related orthogroups.
Indeed, among the mammalian Wnts, 6.6\% of the comparisons in an all-by-all NLE similarity matrix have a `perfect' score of 180 (Figure~\ref{fig:NLEdistributions}), meaning information has been lost that may otherwise have helped differentiate these sequences.
The severity of the problem becomes even more evident when direct comparisons are made between known Wnt orthogroups that are very closely related, such as Wnt2 vs. Wnt2b, Wnt3 vs. Wnt3b, Wnt5a vs. Wnt5b, and Wnt7a vs. Wnt7b, where the proportion of comparisons with NLE values of 180 ranges from 69.6\% to 85.0\% (Figure~\ref{fig:NLEdistributions}).


\begin{figure*}[t]
  \begin{center}
  \includegraphics[height=0.34\textheight]{../figures/NLE_distributions.eps}
\end{center}
\caption{Histograms displaying the distribution of BLASTP-based negative common log of the E-value among mammalian Wnt proteins.
The upper left panel represents all 712,222 comparisons for the 1,194 sequences included in this study, while the remaining panels are restricted to comparisons between orthogroups for each of the seven classes with two named paralogs (e.g., Wnt2/Wnt2b, Wnt3/Wnt3a, Wnt5a/Wnt5b, etc.).}
\label{fig:NLEdistributions}
\end{figure*}


BLAST bit scores are an unconstrained alternative to NLE values, which are used by programs like ProteinOrtho~\cite{Lechner:2011jk} and OrthoFinder~\cite{Emms:2015ig}.
Several issues still remain when switching from E-value to bit score, however, including the direct linear relationship that exists between sequence length and maximum bit scores;
a perfect match between sequences can achieve only about half the score as a perfect match between sequences twice as long (Figure~\ref{fig:Bitscores}A,D).
This leads to reduced recall for orthogroups with short sequences and reduced precision for orthogroups with long sequences~\cite{Emms:2015ig}.
To mitigate this correlation problem, the bit scores can be scaled directly by sequence length (Figure~\ref{fig:Bitscores}B,E) or against a more data-specific linear model as described by Emms and Kelly~\cite{Emms:2015ig} (Figure~\ref{fig:Bitscores}C,F).


\begin{figure*}[t]
  \begin{center}
  \includegraphics[height=0.34\textheight]{../figures/Bitscore_transforms.eps}
\end{center}
\caption{Two-axes histograms displaying the correlation between BLASTP bit score and sequence lengths among mammalian WNT proteins (712,222 comparisons for the 1,194 sequences included in this study).
Darker regions correspond to higher density.
A)~Direct comparison between bit score and length of BLAST alignment (hit length).
A linear relationship exists between the maximum bit score and hit length, so short sequences will be disadvantaged during MCL clustering.
B)~Scaling bit score by hit length equalizes the dynamic range across hit length, but can also inaccurately inflate the implied similarity between sequences which only meet the BLAST similarity threshold along a small localized region.
C)~Total similarity can be better assessed by transforming bit scores against a linear model as described by Emms and Kelly~\cite{Emms:2015ig}. The `Emms transform' was achieved by binning all data according to combined sequence lengths for each pair, selecting the top 5\% from each bin, finding parameters \textit{a} and \textit{b} that best fit the equation log\textsubscript{10}\textit{B\textsubscript{qh}} = \textit{a}log\textsubscript{10}\textit{L\textsubscript{qh}} + \textit{b} (where \textit{B\textsubscript{qh}} is the bitscore and \textit{L\textsubscript{qh}} is the product of the two sequence lengths), then fitting all bit scores to this linear model.
D)~Direct comparison between bit score and full sequence lengths (averaged).
E)~Scaling bit score by full sequence length also equalizes the dynamic range of the scores, but better accounts for cases where short hit lengths in dissimilar sequences over-inflate the scaled score.
F)~The standardizing affect of the Emms transform becomes most noticeable when compared against full sequence length, where both long and short sequences are able to achieve both high and low scores.
}
\label{fig:Bitscores}
\end{figure*}


Despite the improvements provided by the Emms transform, data is still being discarded from the all-by-all similarity matrix when BLAST bit scores are used.
For example, we know \textit{a priori} that all sequences in the WNT dataset are homologous, yet a small percentage of edges are sparse because their similarity falls below the BLAST threshold.
This proportion will become higher when more diverse protein families or greater taxonomic coverage are considered, and while these edges may still be pruned away to optimize orthogroup assignment by MCL (see section~\ref{subsec:TuningParams} below), we cannot know this beforehand.
BLAST is also considering only a single aspect of primary sequence, namely the amino acid substitution matrix (e.g., BLOSUM or PAM).
Of course, an arbitrary number of sequence properties can be included in the similarity metric calculation, so we have devised a new scoring method that forces a global alignment of all sequences and then considers both a substitution matrix and predicted secondary structure properties.
We will refer to this as the SMSS score (\textit{S}ubstitution \textit{M}atrix and \textit{S}econdary \textit{S}tructure).
Figure~\ref{fig:rdmclscores} illustrates how SMSS scores are independent of sequence length and, while correlated with Emms transformed BLASTP bit scores (referred to as the Emms score, from here on), presents a unique dynamic range of values for MCL\@.
The SMSS score is elaborated on in Materials and Methods section~\ref{subsec:seq-sim-func} and it is further compared against the Emms score for orthogroup assignment in section~\ref{subsec:TuningParams}.


\begin{figure}[t]
  \begin{center}
  \includegraphics[height=0.4\textheight]{../figures/rdmcl_scoring.eps}
\end{center}
\caption{Two axes histograms illustrating the relationship between the Substitution Matrix and Secondary Structure (SMSS) scores from paired mammalian WNT sequences and A) the average sequence length; or B) the Emms transformed score.}
\label{fig:rdmclscores}
\end{figure}


\subsection{MCL parameters must be tuned to maximally resolve orthogroups}\label{subsec:TuningParams}
The granularity of MCL is controlled by two parameters \textemdash{} the inflation coefficient and the edge similarity threshold~\cite{VanDongen:kJZ890qx} \textemdash{} and modulating these parameters will change the size and number of the clusters returned.
At present, users of MCL-based orthogroup prediction methods are required to accept predetermined defaults or manually select a small set of static parameters, run MCL, and then choose the outcome they feel best represents their data.
To demonstrate that this is unlikely to yield an optimal set of orthogroups, we randomly sampled 20,000 points in MCL parameter space, using these parameters to cluster the mammalian or avian WNT sequences.
The distribution of cluster number or orthogroups score over parameter space is complex (Figure~\ref{fig:smss_mcl_param_space}) and, while the gross structure is similar, correlation between the mammalian and avian distributions is not perfect (Figure~\ref{fig:mamm_bird_max_scores}).
For example, while the parameters that produce the highest scores from the mammalian dataset also produce relatively high orthogroups scores in the bird dataset, they do not necessarily give the highest scores.
In practice, this means that a single set of optimal MCL parameters for all sequence similarity graphs does not exist,
so the parameters should be dynamically optimized for each unique dataset.


\begin{figure}[t]
  \begin{center}
  \includegraphics[height=0.52\textheight]{../figures/smss_mcl_param_space.eps}
\end{center}
\caption{Scatter plots relating the two MCL parameters (edge similarity threshold and inflation coefficient) to the results of running MCL on either the mammalian WNTs (upper four panels) or the Avian WNTs (lower four panels). MCL output is heat-mapped against either the number of clusters or orthogroups scores and plotted on the z-axes in left-hand panels to best visualize the magnitude of the dynamic range in each dataset. The right-hand panels are a top-down view of the same data, which more clearly illustrates the complex distribution of MCL outputs through parameter space.}
\label{fig:smss_mcl_param_space}
\end{figure}


\begin{figure}[t]
  \begin{center}
  \includegraphics[height=0.25\textheight]{../figures/mamm_bird_max_scores.eps}
\end{center}
\caption{Scatter plot comparing orthogroups scores for mammalian and avian WNTs clustered with the same MCL parameters. The two groups of data points represent the top scoring 5\% of 20,000 randomly sampled parameters from either set of sequences. The correlation between orthogroups scores for these samples is extremely weak.}
\label{fig:mamm_bird_max_scores}
\end{figure}

Given the relatively narrow window of optimal MCL parameters, high-coverage random sampling of parameter space is unnecessary.
However, local maxima exist in the topology of the orthogroups score distribution (Figure~\ref{fig:local_maxima}, so a simple hill-climbing optimization algorithm may also be unideal.
Instead, RD-MCL implements Metropolis coupled Markov chain Monte Carlo~\cite{Hastings:1970iw} (MCMCMC) with simulated annealing~\cite{Kirkpatrick:1983kv}.
When compared against random sampling, using an MCMCMC random-walk heuristic to identify parameters tends to produce higher orthogroups scores from fewer MCL runs (Figure~\ref{fig:random_vs_mcmcmc}).


\begin{figure}[t]
  \begin{center}
  \includegraphics[height=0.4\textheight]{../figures/local_maxima.eps}
\end{center}
\caption{Local orthogroups score maxima exist within MCL parameter space. Arrows point to two such maxim in each scatter plot. (I think this will be fine with only the bottom scatter plot, and maybe increase the density of sampled points)}
\label{fig:local_maxima}
\end{figure}


\begin{figure}[t]
  \begin{center}
  \includegraphics[height=0.5\textheight]{../figures/random_vs_mcmcmc.eps}
\end{center}
\caption{Using an optimization algorithm improves the chance of identifying MCL parameters that maximize orthogroups score. Parameter space was sampled twenty independent times for the mammal and avian WNTs either randomly (2000 points), with three-chain MCMCMC (335 steps on each chain, for 1005 points), or three-chain MCMCMC with simulated annealing (also 335 steps, for 1005 points).}
\label{fig:random_vs_mcmcmc}
\end{figure}


At its core, a high orthogroups score only indicates that a diverse range of taxa exist within the clusters being assessed.
To determine whether we are retrieving correct orthogroups, instead of simply creating cluster diversity, we calculated the recall, precision, true negative (TN) rate, accuracy, and F\textsubscript{1} score for MCL results compared against orthogroups clustered by RefSeq name (Figure~\ref{fig:score_to_accuracy}).
Strong positive correlations exist between all of the metrics and orthogroups scores, except recall.
Recall is complicated by the fact that a small number of large orthogroups will have a high number of true positives along with many false positives, so the measure only becomes useful when combined into a harmonic average with precision (i.e., the F\textsubscript{1} score).
Furthermore, true negatives have a large effect on the calculation of both TN rate and accuracy, so these metrics tend to be quite similar.
We will, therefore, only report F\textsubscript{1} score and accuracy from here on.

\begin{figure*}[t]
  \begin{center}
  \includegraphics[height=0.33\textheight]{../figures/score_to_accuracy.eps}
\end{center}
\caption{Scatter plots showing the positive relationship between orthogroups score and several quality scores (recall, precision, true negative rate, accuracy and F\textsubscript{1} score). Each point represents the clusters of mammalian of avian WNT proteins generated from a randomly selected set of MCL parameters, using the names present in RefSeq as the benchmark for `True' orthogroups.}
\label{fig:score_to_accuracy}
\end{figure*}


Following optimization of MCL parameters, orthogroups derived from graphs of SMSS scores typically outperform those derived from graphs of Emms scores (Figure~\ref{fig:emms_vs_smss}).
It should be clearly noted, however, that this is still a heuristic method with no guarantee of finding the true optimum.


\begin{figure*}[t]
  \begin{center}
  \includegraphics[height=0.25\textheight]{../figures/emms_vs_smss.eps}
\end{center}
\caption{Direct performance comparison between SMSS scores and Emms scores as the similarity metric for MCL. All-by-all graphs of each metric were created for both mammalian and avian WNTs, then MCL parameters were dynamically optimized for each. The best sets of orthogroups from each of 20 replicates were compared against the `true' orthogroups, revealing an average increase in both accuracy and F\textsubscript{1} score when SMSS scores are used as the similarity metric.}
\label{fig:emms_vs_smss}
\end{figure*}


\subsection{Recursive MCL compensates for variable evolutionary separation within and between orthogroups}\label{subsec:recursiveMclCompensatesForVariableEvolutionarySeparationBetweenOrthogroups}
High accuracy

Using the fish as an example group, because there are clearly duplications in the data.
Need to manually tease the orthgroups apart using phylo-trees of each subtype.

\subsection{Number of sequences}
WNTs

Simulated: To test the effects of increasing the number of orthogroups and orthogroup size, datasets were simulated with 4 to 30 taxa and between 4 and 30 genes per taxa.


\subsection{Missing data}
WNTs


\subsection{Gene duplications}
WNTs


\subsection{Hybrid sequences (weird domain structures)}
TIR-containing proteins


\subsection{Simulation data across the dynamic range of RD-MCL}\label{subsec:simulationDataAcrossTheDynamicRangeOfRd-mcl}
To test the performance of RD-MCL compared to other available ortholog prediction tools, we simulated sets of homologs using the Pyvolve module~\cite{Spielman:2015kv}.
These simulations varied in the number of sequences, branch length (substitutions per site), degree of gene loss or duplication, and domain architecture.
The initial seed sequence for all simulations was a polypeptide 398 amino acids long containing four transmembrane domains.
More extensive descriptions of the following simulations can be found in the Methods section.

OrthoFinder performs extremely poorly, but this is probably due to the scaling factor that it generates when classifying orthogroups.
We are not providing enough data to generate a productive linear model~\cite{Emms:2015ig}.
Is it worth while to create a ven diagram illustrating the overlap between methods?


\subsubsection{Branch lengths}
Given an idealized set of homologous sequences, where there has been no gene loss and all gene duplications occurred prior to the last common ancestor of the taxa included in the set, the phylogenetic relationship within each orthogroup should closely approximate the underlying species tree.
Furthermore, the phylogenetic relationship among the orthogroups will approximate the original gene tree of all paralogs present in the last common ancestor.
As such, two distinct axes of divergence must be accounted for when assessing the effect of branch length (i.e., substitutions per site), which we will refer to as the 'species tree length' and 'gene tree length', respectively.

% Maybe make a figure here, to show the two axes of divergence??

A total of 625 datasets were simulated, with each containing eight taxa and seven orthologs (for 56 sequences per dataset).
Branch lengths were varied from 0.05 to 1.55 substitutions per site with standard deviations between 0.05 and 1.05 to prevent perfectly symetrical trees.
As illustrated in Figure~\ref{fig:branch_len_3d}, RD-MCL was either equivalent to or out performed OrthoFinder, OrthoMCL, and ProteinOrtho across the entire dynamic range assessed.
All of the methods tested were more sensitive to changes in species tree branch lengths than they were to gene tree branch lengths (i.e., branches within an orthogroup, as opposed to between orthogroups), although both RD-MCL and OrthoMCL performed marginally better on short species trees with the gene trees were longer.


% May_3_2017_orthofinder_branch_lengths/orthofinder_BL.ipynb
% May_3_2017_OrthoMCL_branch_lengths/orthomcl_BL.ipynb
% May_3_2017_proteinortho_branch_lengths/proteinortho_bl.ipynb
% May_3_2017_rdmcl_branch_len/rdmcl_BL.ipynb
\begin{figure}[t]
  \begin{center}
  \includegraphics[height=0.25\textheight]{../figures/branch_len_3D_scatter.eps}
\end{center}
\caption{Effect of branch length on precision of orthogroup prediction.}
\label{fig:branch_len_3d}
\end{figure}


Figure~\ref{fig:branch_len_std} illustrates the case-by-case performance of RD-MCL compared to the other methods by standardizing the relative precision achieved on each dataset against the precision of RD-MCL\@.


% May_03_2017_branch_lengths.ipynb
\begin{figure}[t]
  \begin{center}
  \includegraphics[height=0.22\textheight]{../figures/branch_len_bargraph.eps}
\end{center}
\caption{Effect of branch length on precision of orthogroup prediction.}
\label{fig:branch_len_std}
\end{figure}


\subsection{Downstream analyses of sample gene families}\label{subsec:downstreamAnalysesOfSampleGeneFamilies}
In this section, I will describe some of the interesting evolutionary and systematic discoveries made possible by fine-grained clustering.


\subsubsection{WNTs}
Big focus on the separation of the fish genes, seeing as the birds and mammals already fit pretty nicely with RefSeq annotations.


\subsubsection{TIR-containing proteins}
Try to fit in some of the work by Mike to say something reasonable about invertebrate sequences.


\subsubsection{Caspases}
Caspases are cysteine-dependent aspartyl-specific proteases with key regulatory roles in inflammation, immunity, and tissue homeostasis~\cite{Songane:2018kq,McIlwain:2013iy}.
The complement of chordate caspases has been well described through careful reconstruction of genomic synteny, capturing a rich history of gene duplication and loss in specific clades~\cite{Eckhart:2008gv,Sakamaki:2009fu,Sakamaki:2015kb,Sakata:2007bs}.
Presence of a caspase catalytic region (CASc) was used as the inclusion criteria

Discuss the mess that is Casps-1/4/5/12/13


\subsubsection{Inferring orthogroup `gene-trees' from consensus sequences}
Reducing the phylogenetic history of a multi-domain gene family down to a single bifurcating tree may not capture the complex evolutionary events behind the modern-day complement of sequences.
Domain shuffling can cause different regions within a gene to have very different genomic histories, with rigorous efforts having been made to develop methods for phylogenetic reconciliation.

\subsubsection{Inferring species trees from RD-MCL generated orthogroups}
Maybe (probably) drop this section.
Otherwise, give examples from each sample dataset - WNTs, caspases, and TLRs.


\subsubsection{Phylogenetic placement of new genes into pre-calculated orthogroups}


\subsection{Implementation}\label{subsec:implementation}
The entire RD-MCL pipeline has been implemented in Python and can be executed from a command-line interface that comes bundled with the software.
While Python is generally OS-agnostic, several of the non-Python third-party programs that RD-MCL depends upon are not compatible with Windows-based systems, so at the time of this writing the RD-MCL package is only supported on Mac OSX and Linux-based systems.
Installation is easily achieved through the Python Package Index (PyPI) using the package manager 'pip' or by downloading the software from GitHub and running the included setup.py script.
More detailed installation instructions and dependency details will be kept up to date in the online documentation (https://github.com/biologyguy/RD-MCL/wiki/Installation-Guide).

Creating all-by-all similarity graphs is an O(n\textsuperscript{2}) hard problem, with RD-MCL run times becoming prohibitive when analyzing more than a few hundred sequences on a single machine.
Unfortunately, Python is not generally well adapted for distributing complex operations across a cluster environment through shared memory.
To overcome this challenge, similarity graph creation can be passed off to worker processes controlling completely independent nodes.
This dynamic is achieved by writing job information to an SQLite database.
Worker nodes retrieve this information, further subdivide the job if appropriate, process the graph, and then return the output to the SQLite database where it is picked back up by the master process.
Distributing the work in this fashion allows an arbitrary number of RD-MCL runs to access a common pool of worker nodes, and the size of that worker pool can be reduced or expanded on the fly as demand fluctuates.

To assist in further analysis, a suite of command-line tools are also included with RD-MCL\@.
These programs allow the user to easily visualize and manipulate the inferred clusters, which includes the ability to group the original input sequences by cluster, rename or merge clusters, and create orthogroup `gene trees' from cluster consensus sequences.


\section{Conclusions}\label{sec:conclusions}
It is important to remember that true orthologs are not intrinsically special.
While negative selection is a strong strong evolutionary pressure that tends to preserve function among orthologs, the fate of any given gene is still at the mercy of biological context and the stochastic nature of inheritance.
Therefore, it can be more appropriate to focus on hierarchical similarity among proteins than on their actual genealogy.


\section{Methods}\label{sec:methods}
\subsection{Simulation data}\label{subsec:simulationData}
The performance of each tool was assessed by calculating the precision and recall of the result on simulated data~\cite{Emms:2015ig}.

\begin{gather*}
    Recall = \frac{TP}{TP + FN}\\
    \\
    Precision = \frac{TP}{TP + FP}\\
    \\
    TN rate = \frac{TN}{TN + FP}
    \\
    Accuracy = \frac{TP + TN}{TP + FP + TN + FN}\\
    \\
    F_1\ Score = 2 * \frac{precision * recall}{precision + recall}\\
\end{gather*}

Where TP is True Positive, FP is False Positives, TN is True Negative, and FN is False Negatives.


\subsection{WNT, TIR-containing, and Caspase sequence data}\label{subsec:wnt,Tir-containing,AndCaspaseSequenceData}
The CASc domain from human caspase-2 (NCBI accession NP\_116764.2) and the wnt domain from mouse Wnt-6 (UniProt accession AAA40569.1) were each used as the starting sequence for a three-iteration PSI-BLAST search of the RefSeq database, limiting the results to metazoans (taxid: 33208), an expectation threshold of 0.01, and allowing for the return of up to 20,000 target sequences.
Accession numbers for all target sequences were downloaded directly from the PSI-BLAST results page, then DatabaseBuddy~\cite{Bond:2017bj} was used to filter out isoforms, partial sequences, and low-quality sequences before fetching the GenBank records.
Given the current depth of species-level coverage in RefSeq, five taxonomic ranges were identified as most appropriate for input into RD-MCL: \textit{Actinopterygii}, \textit{Achelosauria}, \textit{Mammalia}, \textit{Diptera}, and \textit{Hymenoptera}.
The transcriptomes of 25 cnidarian species~\cite{Zapata:2015cc}, assembled using TransDecoder.LongOrfs V.3.0.1~\cite{Haas:2013jq}, were also queried for WNT sequences using BLASTP\@.
These sequences were combined with those cnidarian sequences identified in RefSeq for \textit{Cnidaria}-specific RD-MCL analysis.


\subsection{Sequence similarity function}\label{subsec:seq-sim-func}
Sort sequences by size, Clustal Omega alignment with the --pileup option, per-residue BLOSUM62 log-odds ratios for best possible and observed scores then dividing observed by best, average per-residue difference in PSI-PRED prediction values for coils, \textalpha-helices, and \textbeta-sheets.


\subsection{RD-MCL fitness function}\label{subsec:rd-mclFitnessFunction}
Putative orthogroups were assigned a score based on the size and composition of the cluster, as well as the entire population of sequences available.

Let each sequence $s$ be an element of a set $T$ where all sequences come from the same taxon $j$.

\[
T_j = \{s:s \text{ is a gene in } j\}
\]

All sequences are assigned a score $S$, which is scaled against the largest set of sequences, $T^*$, to bound the minimum score at 1.

\begin{gather*}
    T^* = T_j:|T_j| = max(|T|)\\
    \\
    S_j = \frac{|T^*|}{|T_j|}\\
\end{gather*}

Doing so gives greater weight to those species which have not experienced additional gene expansion, thus allowing greater inclusion of paralogs from those species where gene expansion has been more common.

To penalize the inclusion of paralogs in a putative orthogroup $O$, a diminishing returns algorithm was implemented.
Sequences in the cluster are first sorted into the fewest number of subsets, of largest possible size, where each taxon is represented only once.
This can be expressed as a matrix of size $X \times Y$, where $X$ is the total number of unique taxa and $Y$ is the largest number of sequences derived from a single taxon in the given set.
Each column therefore represents a taxon and is filled from the top down with $S_j$ for each gene it contains, followed by zeros.
For example:


\[
O \equiv
\begin{bmatrix}
    S_{j_1} & S_{j_2} & S_{j_3} & S_{j_4} & S_{j_5} & 0 & S_{j_7}\\
    0 & S_{j_2} & 0 & S_{j_4} & 0 & 0 & S_{j_7} \\
    0 & S_{j_2} & 0 & S_{j_4} & 0 & 0 & 0 \\
    0 & S_{j_2} & 0 & 0 & 0 & 0 & 0 \\
\end{bmatrix}
\]

Each row $Y$ is summed and modified by cofactors $\psi$ and $\gamma$. $\psi$ is proportional to the number of taxa in $Y$ relative to the total number of taxa present globally (i.e.\ the length of $X$ in the above matrix), and $\gamma$ imposes exponentially diminishing returns on the score for each successive index of $Y$.

\begin{gather*}
    \psi = \frac{|\{Y:Y \neq 0\}|}{|j|} + 1\\
    \\
    \gamma = DRB^{Y_{index}}\\
    \\
    S_Y = \gamma\psi\sum_{j} S_j\\
\end{gather*}

Where:

\[
DRB = \text{Diminishing returns base}; 0 \leq DRB \leq 1
\]

The effects of altering $DRB$ are summarized in Figure~\ref{fig:dim_rets}, and we have empirically determined that values between 0.75 and 0.85 generally perform the best.

The final fitness score assigned to a putative orthogroup is thus the sum of each row score:

\[
S_O = \sum_{Y} S_Y
\]

% Apr_19_17_diminishing_returns/dim_returns3D.ipynb
\begin{figure}[t]
  \begin{center}
  \includegraphics[height=0.22\textheight]{../figures/dim_ret_precision_line.eps}
\end{center}
\caption{Effect of diminishing returns base on precision: Doing some stuff with the DRB across sim data (branch length set).}
\label{fig:dim_rets}
\end{figure}


\subsection{Markov chain convergence}\label{subsec:markovChainConvergence}


\subsection{Orphan placement}\label{subsec:orphanPlacement}
\begin{enumerate}
  \item Hidden Markov models are created for each individual sequence and for alignments from each cluster
  \item The probability of emitting each sequence from each sequence hmm is calculated \(forward score\)
  \item Calculate all-by-all pairwise correlation coefficients among sequences, against forward scores
  \item Create a null model from all of the R\textasciicircum2 values from pairs of sequences coming from the same RD-MCL clusters
  \item Create a truncated normal distribution between 0 and 1 for the null model
    \begin{itemize}
    \item I.e., probability that value X is randomly drawn from the population of sequence pair R\textasciicircum2 values in all clusters
    \end{itemize}
  \item Calculate the minimum R\textasciicircum2 value that will be accepted (5\%? 1\%? Use the scipy.ppf() function)
  \item
\end{enumerate}


% \subsection*{Sub-heading for section}
% Text for this sub-heading \ldots
% \subsubsection*{Sub-sub heading for section}
% Text for this sub-sub-heading \ldots
% \paragraph*{Sub-sub-sub heading for section}
% Text for this sub-sub-sub-heading \ldots
% (also see~\cite{koon,khar,zvai,xjon,marg}).
% \nocite{oreg,schn,pond,smith,marg,hunn,advi,koha,mouse}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%%          Backmatter begins here          %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{backmatter}

\section*{Competing interests}
  The authors declare that they have no competing interests.


\section*{Author's contributions}
  SRB is the lead developer of RD-MCL, performed all analyses unless otherwise stated, and wrote the manuscript;
  PG performed the WNT analysis;
  KEK contributed to the code base;
  and ADB was involved in the design and coordination of the project.
  All authors read and approved the final manuscript.


\section*{Acknowledgements}
  This research was supported by the Intramural Research Program of the National Human Genome Research Institute, National Institutes of Health.


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%
%%  Bmc_mathpys.bst  will be used to                       %%
%%  create a .BBL file for submission.                     %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %%
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %%
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% if your bibliography is in bibtex format, use those commands:
\bibliographystyle{bmc-mathphys} % Style BST file (bmc-mathphys, vancouver, spbasic).
\bibliography{../references/refs}      % Bibliography file (usually '*.bib' )
% for author-year bibliography (bmc-mathphys or spbasic)
% a) write to bib file (bmc-mathphys only)
% @settings{label, options="nameyear"}
% b) uncomment next line
%\nocite{label}

% or include bibliography directly:
% \begin{thebibliography}
% \bibitem{b1}
% \end{thebibliography}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Figures                       %%
%%                               %%
%% NB: this is for captions and  %%
%% Titles. All graphics must be  %%
%% submitted separately and NOT  %%
%% included in the Tex document  %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%
%% Do not use \listoffigures as most will included as separate files

%\section*{Figures}
% NOTE: to get the figure to stretch across both columns, use {figure*}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Tables                        %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Use of \listoftables is discouraged.
%%
%\section*{Tables}

%\begin{table}[ht!]
%\caption{List of optional third party software that BuddySuite programs can interact with. BuddySuite performs all necessary format conversion to call any of these tools and, where appropriate, returns the result in the same format as the input. This is particularly useful when creating multiple sequence alignments from annotated sequences in GenBank or EMBL format.}
%      \begin{tabular}{lll}
%        \hline \\
%	   BuddySuite program	& Third-party program					& Reference  \\ \\
%        \hline
%        SeqBuddy			& BLAST 								&~\cite{Chiba:2015ed} \\
%        \hline
%        AlignBuddy			& Clustal Omega 						&~\cite{Chiba:2015ed} \\
%        					& ClustalW2 							&~\cite{Chiba:2015ed} \\
%							& MAFFT 								&~\cite{Chiba:2015ed} \\
%							& MUSCLE 								&~\cite{Chiba:2015ed} \\
%							& PAGAN 								&~\cite{Chiba:2015ed} \\
%        					& PRANK 								&~\cite{Chiba:2015ed} \\
%        \hline
%        PhyloBuddy			& FastTree 								&~\cite{Chiba:2015ed} \\
%        					& RAxML 								&~\cite{Chiba:2015ed} \\
%        					& PhyML 								&~\cite{Chiba:2015ed} \\
%        \hline
%      \end{tabular}
%\label{table:software}
%\end{table}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Additional Files              %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%\section*{Additional Files}
%  \subsection*{Additional file 1 --- Sample additional file title}
%    Additional file descriptions text (including details of how to
%    view the file, if it is in a non-standard format or the file extension).  This might
%    refer to a multi-page table or a figure.

%  \subsection*{Additional file 2 --- Sample additional file title}
%    Additional file descriptions text.


\end{backmatter}
\end{document}
