%% BioMed_Central_Tex_Template_v1.06
%%                                      %
%  bmc_article.tex            ver: 1.06 %
%                                       %

%%IMPORTANT: do not delete the first line of this template
%%It must be present to enable the BMC Submission system to
%%recognise this template!!

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%          <8 June 2012>              %%
%%                                     %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.html and the instructions for   %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%% additional documentclass options:
%  [doublespacing]
%  [linenumbers]   - put the line numbers on margins

%%% loading packages, author definitions

\documentclass[twocolumn]{bmcart}% uncomment this for twocolumn layout and comment line below
%\documentclass{bmcart}

%%% Load packages
%\usepackage{amsthm,amsmath}
%\RequirePackage{natbib}
%\RequirePackage[authoryear]{natbib}% uncomment this for author-year bibliography
%\RequirePackage{hyperref}
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %%
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %%
%%  submitted article.                         %%
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\def\includegraphic{}
\def\includegraphics{}



%%% Put your definitions there:
\startlocaldefs
\endlocaldefs


%%% Begin ...
\begin{document}

%%% Start of article front matter
\begin{frontmatter}

\begin{fmbox}
\dochead{Software}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the title of your article here     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\title{Recursive dynamic Markov clustering for fine-grained orthogroup classification}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors here                   %%
%%                                          %%
%% Specify information, if available,       %%
%% in the form:                             %%
%%   <key>={<id1>,<id2>}                    %%
%%   <key>=                                 %%
%% Comment or delete the keys which are     %%
%% not used. Repeat \author command as much %%
%% as required.                             %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\author[
   addressref={aff1},                    % id's of addresses, e.g. {aff1,aff2}
   %noteref={n1},                        % id's of article notes, if any
   %corref={aff1},                       % id of corresponding address, if any
   %email={steve.bond@nih.gov}            % email address
]{\inits{SR}\fnm{Stephen R} \snm{Bond}}
\author[
   addressref={aff1},
   %email={karlkeat@gmail.com}
]{\inits{KE}\fnm{Karl E} \snm{Keat}}
\author[
   addressref={aff1},
   email={andy@mail.nih.gov}   
]{\inits{AD}\fnm{Andreas D} \snm{Baxevanis}}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors' addresses here        %%
%%                                          %%
%% Repeat \address commands as much as      %%
%% required.                                %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\address[id=aff1]{%                           % unique id
  \orgname{Computational and Statistical
   Genomics Branch, Division of Intramural    % university, etc
    Research, National Human Genome
     Research Institute, National
      Institutes of Health},
  \street{50 South Drive},                     %
  \postcode{20892}                             % post or zip code
  \city{Bethesda},                             % city
  \state{MD},
  \cny{USA}                                    % country
}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter short notes here                   %%
%%                                          %%
%% Short notes will be after addresses      %%
%% on first page.                           %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%\begin{artnotes}
%\note{Sample of title note}     % note to the article
%\note[id=n1]{Equal contributor} % note, connected to author
%\end{artnotes}

%\end{fmbox}% comment this for two column layout

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Abstract begins here                 %%
%%                                          %%
%% Please refer to the Instructions for     %%
%% authors on http://www.biomedcentral.com  %%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{abstractbox}

\begin{abstract} % abstract
\parttitle{Background}  % Required
Blahh

\parttitle{Results}  % Required
Blahh

\parttitle{Conclusions}  % Required
Blahh

\end{abstract}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The keywords begin here                  %%
%%                                          %%
%% Put each keyword in separate \kwd{}.     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{keyword}  % Three to ten keywords
\kwd{orthogroup}
\kwd{ortholog}
\kwd{Markov clustering}
\end{keyword}

% MSC classifications codes, if any
%\begin{keyword}[class=AMS]
%\kwd[Primary ]{}
%\kwd{}
%\kwd[; secondary ]{}
%\end{keyword}

\end{abstractbox}
%
\end{fmbox} % uncomment this for twcolumn layout

\end{frontmatter}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Main Body begins here                %%
%%                                          %%
%% Please refer to the instructions for     %%
%% authors on:                              %%
%% http://www.biomedcentral.com/info/authors%%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%% See the Results and Discussion section   %%
%% for details on how to create sub-sections%%
%%                                          %%
%% use \cite{...} to cite references        %%
%%  \cite{koon} and                         %%
%%  \cite{oreg,khar,zvai,xjon,schn,pond}    %%
%%  \nocite{smith,marg,hunn,advi,koha,mouse}%%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%% start of article main body
% <put your article body there>

%%%%%%%%%%%%%%%%
%% Background %%
%%
\section*{Background and rationale}
When a gene evolves an important physiological function, purifying selection tends to maintain that function through evolutionary time \cite{Altenhoff:2012ea, Rogozin:2014fp, KryuchkovaMostacci:2016iw}. As a result, orthology (i.e., homology via speciation) has become a widely used predictor of shared gene product function among species, with considerable effort made to develop computational methods for identifying orthologs. The algorithms currently in popular use fall into two broad categories: Tree-based and graph-based clustering methods (recently reviewed by Fredj Tekaia \cite{Tekaia:2016ga}). Breifly, tree-based approaches (e.g., Ensembl Compara \cite{Vilella:2009ju}, LOFT \cite{vanderHeijden:2007bo}, and SYNERGY \cite{Wapinski:2007fa}) identify orthologous clades by estimating phylogenetic trees for a target gene family, and then attempt to reconcile those gene trees against a `known' species tree. While tree-based methods are very accurate under ideal conditions, they are very sensitive to the accuracy of the species trees they rely on, which can become a considerable source of uncertainty or error \cite{Xu:2016ek}. Alternatively, pairwise similarity graph clustering methods leverage graph theory to rapidly identify groups of related sequences from genome scale datasets. Due to the non-transitive nature of orthology (i.e., paralogs in one species can be orthologous to a single gene in another species), groupings of pure orthologs may not be possible. Instead, the term `orthogroup' has come to represent a cluster of genes desended from a common ancestor of the clade in question, which may include paralogs \cite{Wapinski:2007fa}. InParanoid \cite{OBrien:2005cy}, EggNOG \cite{Jensen:2007cc}, and OMA \cite{Roth:2009iu} are popular tools for assigning sequences to orthogroups using a `best-hit clique' approach, where closed best-hit sub-graphs are identified in the dataset. While accurate within each sub-graph, these methods tend to be overly strict in their assignment; this causes an under-representation of actual orthologous relationships among many species. Alternatively, Markov clustering (MCL) is very efficient at isolating more inclusive sub-graphs. OrthoMCL is one of the most popular MCL-based ortholog prediction methods \cite{Li:2003en}, but it is prone to placing too many in-paralogs into orthogroups (i.e., it is less precise).

For coarse-grained, genome-wide analysis, many of the tools mentioned above perform very well.

In the current study we have increased the overall resolving power of MCL-based orthogroup assignment with a number of novel enhancements, including refinement of the pairwise similarity metrics, using an optimization algorithm to dynamically select MCL parameters, recursively subdividing orthogroups, and testing putative orthogroups for best-hit cliques to maximize resolution.


\section*{Results}
The imputus for developing RD-MCL was to predict high-quality fine-grained orthogroups among sequences from a defined gene family.

\subsection*{Description of the RD-MCL algorithm and software}
Spill some ink regarding in/out paralogs \cite{Sonnhammer:2002vm,Tekaia:2016ga}.

BLAST scores (bit or e-value) have a strong length bias when calculating orthogroups \cite{Emms:2015ig}. OrthoFinder also uses a static inflation/edge similarity threashold \cite{Emms:2015ig}

\subsection*{Simulation data across the dynamic range of RD-MCL}
\subsubsection*{Branch lengths}
Ipsum lorum

\subsubsection*{Number of sequences}
Ipsum lorum

\subsubsection*{Missing data}
Ipsum lorum

\subsubsection*{Gene duplications}
Ipsum lorum

\subsubsection*{Hybrid sequences (weird domain structures)}
Ipsum lorum

\subsection*{RD-MCL classification of known gene families}
Ipsum lorum

\subsubsection*{Gene family 1}
Ipsum lorum

\subsubsection*{Gene family 2}
Ipsum lorum

\subsection*{RD-MCL classification of new gene families}
Ipsum lorum

\subsubsection*{Gene family 1}
Ipsum lorum

\subsubsection*{Gene family 2}
Ipsum lorum

\section*{Conclusions}
Can't wait to share my conclusions.

\section*{Methods}
\subsubsection*{RD-MCL fitness function}
Putative orthogroups were assigned a score based on the size and composition of the cluster, as well as the entire population of sequences available.

Let each sequence $s$ be an element of a set $T$ where all sequences come from the same taxa $j$.

\[ T_j = \{s:s \in j\} \]

All sequences are assigned a base score $S$, and the minimum score is set to 1. This minimum score is applied to all sequences belonging to the largest set of sequences $T^*$.

\[ T^* = T_j:|T_j| = max(|T|) \]

\[S_{T^*} = 1 \]

All remaining squences are assigned a value greater than 1, proportional to the size of the set it belongs to relative to $T^*$.

\[ S_{T_j} = \frac{|T^*|}{|T_j|} \]


To penalize the inclusion of multiple sequences from a single taxa in a given putative orthogroup $O$, a diminishing returns function was implemented; whereby the sequences in the cluster were sorted into the fewest number of subsets of largest possible size where each taxa is represented only once. The sum score for each subset is determined, and then in diminishing order of subset size,

\subsubsection*{Markov chain convergence}


- I used MAFFT \cite{Katoh:2013hm}, because it's awesome.

- The COG, KOG, arCOG databases may all be rich sources of data for validation. COGs are `clusters of orthologous genes', which can includes many individual orthogroups.

- KEGG OCs may work in place of KOGs \cite{Nakaya:2013gg}

- Construct species trees with *BEAST (Heled and Drummond 2010) and BPP (Yang and Rannala 2014; Rannala and Yang 2016) to do a tree based comparison against RD-MCL.

- Possible sample data: CYP proteins \cite{Pan:2016jg}

- Might want to compare results agains Ortholog-Finder if appropriate \cite{Horiike:2016dq}

- Construction of an ortholog ontology \cite{Chiba:2015ed}?

- Try OrthoFinder length-normalized bit scores as similarity metric between sequences \cite{Emms:2015ig}.

- Use precision and recall as measures of accuracy of simulated data

(https://en.wikipedia.org/wiki/Precision\_and\_recall and \cite{Emms:2015ig})

$Precision = \frac{TP}{TP + FP}$

$Recall = \frac{TP}{TP + FN}$

$F-Score = 2 * \frac{precision * recall}{precision + recall}$

Where TP is True Positive, FP is False Positives, and FN is False Negatives.

- Create raw clusters using MMseqs2.0

(https://github.com/soedinglab/MMseqs2).

- Output orthogroups in plain text and the \textit{Quest for Orthologs} community standard OrthoXML \cite{Dessimoz:2012dj}


% \subsection*{Sub-heading for section}
% Text for this sub-heading \ldots
% \subsubsection*{Sub-sub heading for section}
% Text for this sub-sub-heading \ldots
% \paragraph*{Sub-sub-sub heading for section}
% Text for this sub-sub-sub-heading \ldots
% (also see \cite{koon,khar,zvai,xjon,marg}).
% \nocite{oreg,schn,pond,smith,marg,hunn,advi,koha,mouse}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Backmatter begins here                   %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{backmatter}

\section*{Competing interests}
  The authors declare that they have no competing interests.

\section*{Author's contributions}
  SRB is the lead developer of RD-MCL and wrote the manuscript, KEK contributed significantly to the code base, and ADB was involved in the design and coordination of the project. All authors read and approved the final manuscript.

\section*{Acknowledgements}
  This research was supported by the Intramural Research Program of the National Human Genome Research Institute, National Institutes of Health.
  

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%
%%  Bmc_mathpys.bst  will be used to                       %%
%%  create a .BBL file for submission.                     %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %%
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %%
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% if your bibliography is in bibtex format, use those commands:
\bibliographystyle{bmc-mathphys} % Style BST file (bmc-mathphys, vancouver, spbasic).
\bibliography{../references/refs}      % Bibliography file (usually '*.bib' )
% for author-year bibliography (bmc-mathphys or spbasic)
% a) write to bib file (bmc-mathphys only)
% @settings{label, options="nameyear"}
% b) uncomment next line
%\nocite{label}

% or include bibliography directly:
% \begin{thebibliography}
% \bibitem{b1}
% \end{thebibliography}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Figures                       %%
%%                               %%
%% NB: this is for captions and  %%
%% Titles. All graphics must be  %%
%% submitted separately and NOT  %%
%% included in the Tex document  %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%
%% Do not use \listoffigures as most will included as separate files

\section*{Figures}
  \begin{figure}[h!]
  \caption{\csentence{Sample figure title.}
      A short description of the figure content
      should go here.}
      \end{figure}

\begin{figure}[h!]
  \caption{\csentence{Sample figure title.}
      Figure legend text.}
      \end{figure}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Tables                        %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% Use of \listoftables is discouraged.
%%
\section*{Tables}

\begin{table}[h!]
\caption{List of optional third party software that BuddySuite programs can interact with. BuddySuite performs all necessary format conversion to call any of these tools and, where appropriate, returns the result in the same format as the input. This is particularly useful when creating multiple sequence alignments from annotated sequences in GenBank or EMBL format.}
      \begin{tabular}{lll}
        \hline \\
	   BuddySuite program	& Third-party program					& Reference  \\ \\
        \hline
        SeqBuddy			& BLAST 								& \cite{Camacho2009} \\
        \hline
        AlignBuddy			& Clustal Omega 						& \cite{Sievers:2011fn} \\
        					& ClustalW2 							& \cite{Larkin:2007hz} \\
							& MAFFT 								& \cite{Katoh:2013hm} \\
							& MUSCLE 								& \cite{Edgar:2004bo} \\
							& PAGAN 								& \cite{Loytynoja:2012fy} \\
        					& PRANK 								& \cite{Loytynoja:2005cb} \\				
        \hline
        PhyloBuddy			& FastTree 								& \cite{Price:2010eg} \\
        					& RAxML 								& \cite{Stamatakis:2006de} \\
        					& PhyML 								& \cite{Guindon:2010gf} \\
        \hline
      \end{tabular}
\label{table:software}
\end{table}

\begin{table}[h!]
\caption{File format support for reading (R) and writing (W) provided by each BuddySuite module.}
      \begin{tabular}{llll}
        \hline \\
        Format							& SeqBuddy  					& AlignBuddy   					& PhyloBuddy\\ 
        \\
        \hline
        Clustal 						& R \& W\textsuperscript{\dag} 	& R \& W						& None \\ 
        EMBL\textsuperscript{\ddag} 	& R \& W						& R\textsuperscript{\dag}/ W	& None \\
        FASTA 							& R \& W						& R\textsuperscript{\dag}/ W	& None \\
        GenBank\textsuperscript{\ddag} 	& R \& W						& R\textsuperscript{\dag}/ W 	& None \\
        Nexus 							& R \& W\textsuperscript{\dag}	& R \& W						& R \& W \\ 
        Newick 							& None							& None							& R \& W \\ 
        NeXML							& None							& None							& R \& W \\
        PHYLIP (interleaved)			& R \& W\textsuperscript{\dag} 	& R \& W						& None \\
        PHYLIP (sequential)				& R \& W\textsuperscript{\dag} 	& R \& W						& None \\	
        SeqXML							& R \& W						& None							& None \\ 
        Stockholm						& R \& W\textsuperscript{\dag} 	& R \& W						& None \\ 
        Swissprot\textsuperscript{\ddag}& R only						& None							& None \\
        \hline
      \end{tabular}
      \begin{tabular}{l}
      \textsuperscript{\dag}All sequences must be the same length \\
      \textsuperscript{\ddag}Supports rich sequence annotation \\
      \hline
       \end{tabular}
\label{table:formats}
\end{table}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Additional Files              %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\section*{Additional Files}
  \subsection*{Additional file 1 --- Sample additional file title}
    Additional file descriptions text (including details of how to
    view the file, if it is in a non-standard format or the file extension).  This might
    refer to a multi-page table or a figure.

  \subsection*{Additional file 2 --- Sample additional file title}
    Additional file descriptions text.


\end{backmatter}
\end{document}
